separate out diloco configs

tushar00jain opened the pull request.

tianyu-l commented on torchtitan/components/ft/config/job_config.py:
```diff
@@ -0,0 +1,80 @@
+
+
+@dataclass
+class FaultTolerance(BaseFaultTolerance):
```
I don't think you need to subclass, per https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/manager.py#L123

Have you tried and it didn't work?

tushar00jain replied:
yes, don't get types

tianyu-l replied:
We were able to do this for the FLUX.1 model https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/job_config.py#L11

Is the obstacle around typing? If so I think we could bear with imperfect typings for now, for the sake of minimality, as the repo has other places which may break typing anyway (which we should continue improving).

tushar00jain replied:
nvm, i was able to use getattr

tianyu-l commented on torchtitan/components/ft/config/job_config.py:
```diff
@@ -0,0 +1,80 @@
+    """
+
+
+@dataclass
+class JobConfig:
+    fault_tolerance: FaultTolerance = field(default_factory=FaultTolerance)
```
without subclassing, this is not needed

tushar00jain replied:
FaultTolerance is subclassed

tianyu-l commented on torchtitan/components/ft/diloco.py:
I suggest calling this file `torchtitan/components/ft/diloco/utils.py`

tianyu-l commented on torchtitan/components/ft/protocol/model.py:
can be put in `torchtitan/components/ft/diloco/__init__.py`

tianyu-l commented on torchtitan/components/ft/protocol/model.py:
```diff
@@ -0,0 +1,24 @@
+
+
+@dataclass
+class FaultTolerantTrainSpec(TrainSpec):
```
maybe `SemiSyncTrainSpec` or `DiLoCoTrainSpec` to be not confused with other FT features, but OK either way.

tianyu-l commented on torchtitan/train.py:
```diff
@@ -48,6 +52,7 @@ class Trainer(torch.distributed.checkpoint.stateful.Stateful):
     lr_schedulers: train_spec_module.LRSchedulersContainer
     validator: train_spec_module.BaseValidator
     metrics_processor: train_spec_module.MetricsProcessor
+    model_args: train_spec_module.BaseModelArgs
```
please respect order ..

tianyu-l commented on torchtitan/train.py:
```diff
@@ -17,6 +17,10 @@
 from torchtitan.components.checkpoint import CheckpointManager
 from torchtitan.components.dataloader import DataloaderStopIteration
 from torchtitan.components.ft import FTManager, maybe_semi_sync_training
+from torchtitan.components.ft.protocol import (
+    FaultTolerantModelArgs,
+    FaultTolerantTrainSpec,
+)
```
pls remove these imports

tushar00jain replied:
removed, used getattr

tianyu-l commented on torchtitan/train.py:
```diff
@@ -545,8 +551,18 @@ def train(self):
                 ft_manager=self.ft_manager,
-                model_parts=self.model_parts,
+                model=self.model_parts[0],
+                n_layers=(
+                    self.model_args.n_layers
+                    if isinstance(self.model_args, FaultTolerantModelArgs)
+                    else 0
+                ),
```
let's just pass in `model_args`

tushar00jain replied:
can't, it creates circular dependency because TrainSpec depends on ft manager

tianyu-l replied:
Could you clarify the circular dependency? It seems `TrainSpec` and `FaultTolerantTrainSpec` don't depend on on `FTManager`.

tushar00jain replied:
it does. through


in `train_spec.py` -- `from torchtitan.components.lr_scheduler import LRSchedulersContainer`
in  `lr_scheduler.py` -- `from torchtitan.components.optimizer import OptimizersContainer`

tushar00jain replied:
the impots are not needed anymore using getattr

tianyu-l commented on torchtitan/train.py:
```diff
@@ -545,8 +551,18 @@ def train(self):
+                    else 0
+                ),
                 optimizer=self.optimizers,
+                fragment_fn=(
```
can we pass in train_spec?

tushar00jain replied:
can't also creates circular dependency, btw this change is on the previous pr

tushar00jain replied:
don't need this anymore

tianyu-l commented on torchtitan/models/llama3_ft/__init__.py:
```diff
@@ -0,0 +1,69 @@
+]
+
+
+class FaultTolerantTransformerModelArgs(
```
why do we need this?

tushar00jain replied:
removed as suggested, just using getattr now as suggested

tianyu-l commented on torchtitan/models/llama3_ft/__init__.py:
```diff
@@ -0,0 +1,69 @@
+        super().__init__(*args, **kwargs)
+
+
+llama3_configs = {
```
why do we need this?

tushar00jain replied:
for typing otherwise it won't be a instance of the desired class

tushar00jain replied:
don't need to do this anymore

tianyu-l commented on torchtitan/components/ft/protocol/model.py:
```diff
@@ -0,0 +1,24 @@
+
+
+@dataclass
+class FaultTolerantModelArgs(BaseModelArgs):
```
why do we need this? If it's for typing, you can assert the existence of `n_layers` before use.

tushar00jain replied:
we cant do that on basebodel args? i think we should have a concrete type otherwise usage perspective its pretty implicit

tianyu-l replied:
I think right now they typing is not perfect. E.g. PP uses https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/infra/pipeline.py#L55
We should bias towards simplicity until we have a more correct abstraction. Putting `n_layers` in `FaultTolerantModelArgs` doesn't sound the right abstraction.

tushar00jain replied:
done

tushar00jain commented on torchtitan/components/ft/config/__init__.py:
why overkill if the suggestion is to put this in a different folder? or you mean something else?

tianyu-l reviewed (changes requested):
Thanks for replying. I left some more comments. Sorry if the directions I gave may not make sense to you.

There is a possibility that, compared with the basic FT features that are already shipped in torchtitan, the difficulty of making streaming diloco changes minimal is due to its intrinsic complexity.

If that's the case, we should put streaming diloco under `torchtitan/experiments` folder, partly to unblock important collaborations.

Putting in `experiments` doesn't mean this feature itself is not important or not ready. One reason could be we haven't found a stabilized API, e.g. for other internal work [SimpleFSDP](https://github.com/pytorch/torchtitan/tree/main/torchtitan/experiments/simple_fsdp), [TorchForge](https://github.com/pytorch/torchtitan/tree/main/torchtitan/experiments/forge).

cc @d4l3k @fegin

tushar00jain commented:
@tianyu-l thanks for clarifying that torchtitan the type system in torchtitan isn't perfect yet. we can stick with the current approach if that's the case. i updated the diff accordingly. seems the only thing remaining now is to agree is,

- on the folder structure for https://github.com/pytorch/torchtitan/pull/1446 hopefully after addressing your comments, the changes are kept minimal and we don't need to put this in the experiments folder
- on the subclassing of job config for https://github.com/pytorch/torchtitan/pull/1516 -- lmk if i should also use `hasattr` for this and remove the subclassing. there's a bunch of config options there so maybe subclassing is ok there so that we can use the type in `ft` folder to access all fields, including the merged ones, from `fault_tolerance` config?

tushar00jain commented:
@tianyu-l Thanks for the thorough feedback and making sure it's easy for users to remove features they don't need!
